### ISSUE TITLE
Emscripten cursor position wrong when scaling is not 1

### ISSUE THREAD

**Issue by SirNate0** [urho3d/Urho3D#2768]
I'm creating a web game, and when the page starts with a `window.devicePixelRatio` that is not 1 the UI reacts to the mouse at the wrong position (see image, with 1.25 scaling).
![image](https://user-images.githubusercontent.com/10181610/106365868-51e75200-6306-11eb-903b-cfa148a566c3.png)

Changing the size after loading does not alter it, it still retains the same ratio offset when the scaling is changed after load (so if I go back to 100% zoom after loading the mouse still highlights New Game on hover when it is off to the left and up). Only refreshing the page after resetting the zoom fixes it.

Further confusing the issue, on Windows (and probably some other operating systems) setting a system-wide scaling will make the default 100% zoom in Chrome correspond to a non-1 devicePixelRatio. E.g. with a system scaling of 1.25, 100% zoom for the page corresponds to `window.devicePixelRatio == 1.25`, so you have to zoom to 80% and reload to get the proper behavior.

(Note: I haven't tested with the latest Master, just from Wed Nov 25, 2020, so I can try changing to the latest master, but [Web Sample 37: UI Drag](https://urho3d.github.io/samples/37_UIDrag.html) is still broken, so I doubt that is the issue)

**Comment by SirNate0** [comment]
I haven't thoroughly tested it, but I am finding that just forcing the UI scaling to be 1 seems to fix the issue, something like
```
SubscribeToEvent(E_SCREENMODE,[&](StringHash eventType, VariantMap& eventData) {
    auto ui = GetSubsystem<UI>();
    if (ui)
    {
        ui->SetScale(1);
        URHO3D_LOGINFO("Forcing UI Scale to 1.");
    }
});
```

**Comment by JTippetts** [comment]
Can confirm, the bug appears in my own game. The workaround @SirNate0 posted will work, but if the user changes the zoom level afterward (at least on chrome, where I've tested it) the bug re-occurs. Calling ui->SetScale(1) inside the update cycle solves that, though it seems like needless work to be done every frame.

**Comment by 1vanK** [comment]
https://github.com/urho3d/Urho3D/pull/2678

**Comment by SuperWangKai** [comment]
Found in C++ application.

Problem solved after a quick test (POC), I added these two lines
```c++
    positionCopy.x_ = int(positionCopy.x_ / uiScale_);
    positionCopy.y_ = int(positionCopy.y_ / uiScale_);
```

to `UIElement* UI::GetElementAt(UIElement* root, const IntVector2& position, bool enabledOnly)` before
https://github.com/urho3d/Urho3D/blob/0e488a902d89aa8e9534955313d2b0c103f28cdd/Source/Urho3D/UI/UI.cpp#L842

It seems `GetElementAt` does not take UI scale into caculation.


The working code -
```c++
UIElement* UI::GetElementAt(UIElement* root, const IntVector2& position, bool enabledOnly)
{
    IntVector2 positionCopy(position);
    const IntVector2& rootSize = root->GetSize();
    const IntVector2& rootPos = root->GetPosition();

    // If position is out of bounds of root element return null.
    if (position.x_ < rootPos.x_ || position.x_ > rootPos.x_ + rootSize.x_)
        return nullptr;

    if (position.y_ < rootPos.y_ || position.y_ > rootPos.y_ + rootSize.y_)
        return nullptr;

    // If UI is smaller than the screen, wrap if necessary
    if (rootSize.x_ > 0 && rootSize.y_ > 0)
    {
        if (positionCopy.x_ >= rootPos.x_ + rootSize.x_)
            positionCopy.x_ = rootPos.x_ + ((positionCopy.x_ - rootPos.x_) % rootSize.x_);
        if (positionCopy.y_ >= rootPos.y_ + rootSize.y_)
            positionCopy.y_ = rootPos.y_ + ((positionCopy.y_ - rootPos.y_) % rootSize.y_);
    }

    //IntVector2 pos(eventData[P_X].GetInt(), eventData[P_Y].GetInt());
    positionCopy.x_ = int(positionCopy.x_ / uiScale_);
    positionCopy.y_ = int(positionCopy.y_ / uiScale_);

    UIElement* result = nullptr;
    GetElementAt(result, root, positionCopy, enabledOnly);
    return result;
}

```

**Comment by SuperWangKai** [comment]
However, in "Touch mode", touch position is already corrected, so my code won't work when touch is enabled.

**Comment by dprandle** [comment]
Hi all - there was a fairly recent merge that removed the lines (see 1VanK post above)
`
    pos.x_ = (int)(pos.x_ / uiScale_);
    pos.y_ = (int)(pos.y_ / uiScale_);
`
from the function Ui::GetCursorPositionAndVisible - I mentioned that this broke UI scaling for me and it sounds related to what is happening here..

I did some more investigation - I don't have the solution but maybe some useful info..

The editor (Urho3DPlayor) automatically has it's global UI scaling set to 2.0 on high DPI displays, it does this by checking that the input scaling is not 1... and if not it changes the UI scale to 2, then changes the UI.png to FILTER_NEAREST.. I assume that the engine set's input scale automatically to 2 for high DPI displays.

I tested this with debug output - _uiScale is indeed set to 2.0 along with input scale - both x_ and y_ are set to 2.0 for uhro3dPlayer.

Somewhere, urho player is detecting that im using a high DPI display and setting it as such, because the urho examples don't. Take UIDrag (example 37) - the same log shows the uiScale_ is set to 1.0 while input scaling x_ and y_ are both 2.0.

Strange part.. for me - removing the division by uiScale_ in UI::GetCursorPositionAndVisible makes UI mouse position behave correctly in Urho3DPlaye, adding those lines back in makes it behave _incorrectly_. But, for UIDrag example, removing the lines behaves correctly if the uiScale_ is 1.0 but incorrectly if the uiScale_ is 2.0... but if the division by uiScale_ is included in UI::GetCursorPositionAndVisible then UIDrag behaves correctly whether the uiScale_ is 1.0 or 2.0.
